# 2장 1강: t-검정의 이론과 가정 — 실습문제

## 실습 목표

- 평균 차이와 표준오차를 이용해 t통계량을 직접 계산할 수 있다.
- 단일표본 t검정으로 하나의 표본평균과 기준값을 비교할 수 있다.
- 독립성·정규성·등분산성을 확인한 뒤 독립표본 t검정을 수행할 수 있다.
- 독립표본과 대응표본 상황을 구분하고 적절한 함수를 선택할 수 있다.
- p-value를 유의수준과 비교하여 검정 결과를 올바르게 해석할 수 있다.

## 실습 환경 / 데이터

- Python
- NumPy, pandas
- scipy.stats
- `ames_housing.csv`

| 컬럼 | 의미 |
|---|---|
| `SalePrice` | 주택 판매가격 |
| `KitchenQual` | 주방 품질 |

> 모든 검정은 양측검정이며 유의수준 `α = 0.05`를 사용합니다.  
> Ames 표본 추출에는 지정된 `random_state`를 사용해 결과를 재현합니다.

## 실습 준비

1. 필요한 라이브러리를 불러오세요.
2. Ames Housing 데이터를 `df`에 불러오세요.
3. 데이터 크기, 전체 결측치 수, 컬럼명과 상위 5개 행을 확인하세요.

In [2]:
# 실습 준비 코드를 작성하세요.
import pandas as pd
from scipy import stats

df = pd.read_csv('ames_housing.csv')

df.describe()

df.head()

,SalePrice,GrLivArea,LotArea,OverallQual,KitchenQual,CentralAir,HeatingQC,PavedDrive,Neighborhood,YearBuilt
0,208500,1710,8450,7,Gd,Y,Ex,Y,CollgCr,2003
1,181500,1262,9600,6,TA,Y,Ex,Y,Veenker,1976
2,223500,1786,11250,7,Gd,Y,Ex,Y,CollgCr,2001
3,140000,1717,9550,7,Gd,Y,Gd,Y,Crawfor,1915
4,250000,2198,14260,8,Gd,Y,Ex,Y,NoRidge,2000


In [3]:
# T - 통계량: 관찰된 평균 차이가 표준오차의 몇 배인지를 나타내는 검증용 숫자
# -> 평균이 75점, 기준이 70점, 표준오차가 약 0.57점이면 t값은 약 8.6이고
# 평균의 차이가 표준오차의 약 8.6배 정도 라는 뜻

# 독립표본 t 검정: 서로 독립된 두 집단의 모평균이 같은지 비교하는 검정
# -> 광고 A와 B를 본 서로 다른 고객의 평균 구매금액을 비교한다.

# welch t 검정: 두 집단의 분산이 같다는 가정 X, 독립된 두 집단의 평균을 비교

# Shapiro : 데이터가 정규분포에서 왔다는 귀무가설을 검토하는 정규성 검정
# -> P값이 0.05보다 작으면 정규성 가정과 맞지 않는 근거가 있다고 해석한다. 

# Levene : 여러 집단의 분산이 같다는 귀무가설을 검토하는 검정
# -> 두 집단 비교에서 P값이 0.05보다 작으면 등분산 가정과 맞지 않는 근거가 있다고 해석한다.

---

## 필수 1. t통계량 직접 계산과 단일표본 t검정

### 문제 1-1. 평균 판매가격은 180,000달러와 다른가?

#### 문제 설명

Ames 주택 30개를 표본으로 추출하여 평균 판매가격이 기준값 180,000달러와 다른지 확인합니다. 먼저 t통계량을 직접 계산한 뒤 `ttest_1samp()` 결과와 비교하세요.

#### 요구사항

1. `SalePrice`에서 `n=30`, `random_state=2`로 표본을 추출하여 `sale_sample`에 저장하세요.
2. 표본 수, 표본평균, 표본표준편차를 출력하세요.
3. `stats.sem()`을 이용해 표준오차를 계산하세요.
4. 다음 식으로 t통계량을 직접 계산하세요.  
   `t = (표본평균 - 기준값) / 표준오차`
5. Shapiro-Wilk 검정으로 표본의 정규성을 확인하세요.
6. 다음 가설을 작성하세요.
   - H₀: 모집단 평균 판매가격은 180,000달러이다.
   - H₁: 모집단 평균 판매가격은 180,000달러가 아니다.
7. `stats.ttest_1samp()`로 단일표본 t검정을 수행하세요.
8. 직접 계산한 t통계량과 함수가 반환한 t통계량을 비교하세요.
9. p-value를 이용해 귀무가설 기각 여부를 판단하세요.

#### 해석 질문

**Q1.** t통계량은 평균 차이와 표준오차를 어떻게 이용한 값인가요?  
**Q2.** 같은 평균 차이라면 표준오차가 작아질수록 t통계량의 절댓값은 어떻게 변하나요?  
**Q3.** 직접 계산한 t통계량과 `ttest_1samp()`의 t통계량은 일치하나요?  
**Q4.** 검정 결과 평균 판매가격이 180,000달러와 다르다고 판단할 수 있나요?

#### 제출 결과

- 기술통계량과 정규성 결과
- 직접 계산한 표준오차와 t통계량
- 단일표본 t검정 결과
- 귀무가설 판단과 해석
- Q1~Q4 답변

In [4]:
# 필수 1 코드를 작성하세요.
import numpy as np
import pandas as pd
from scipy import stats

# 1. SalePrice에서 30개 표본 추출
sale_sample = df["SalePrice"].dropna().sample(n=30, random_state=2)

# 2. 표본 수, 표본평균, 표본표준편차(ddof=1)
n = len(sale_sample)
sample_mean = sale_sample.mean()
sample_std = sale_sample.std(ddof=1)

print("=== 1. 기술통계량 ===")
print(f"표본 수 (n)        : {n}")
print(f"표본평균           : ${sample_mean:,.2f}")
print(f"표본표준편차 (s)   : ${sample_std:,.2f}")

# 3. 표준오차(SE) 계산
se = stats.sem(sale_sample)

# 4. t통계량 직접 계산
pop_mean = 180000  # 기준값
t_manual = (sample_mean - pop_mean) / se

print("\n=== 2. 직접 계산 결과 ===")
print(f"표준오차 (SE)      : ${se:,.4f}")
print(f"직접 계산 t통계량  : {t_manual:.4f}")

# 5. Shapiro-Wilk 정규성 검정
stat_shapiro, p_shapiro = stats.shapiro(sale_sample)
print("\n=== 3. 정규성 검정 ===")
print(f"Shapiro-Wilk 통계량: {stat_shapiro:.4f}, p-value: {p_shapiro:.4f}")
print(f"정규성 충족 여부   : {'충족' if p_shapiro >= 0.05 else '미충족'}")

# 6 & 7. 단일표본 t검정 수행 (양측검정)
t_stat, p_value = stats.ttest_1samp(
    sale_sample, popmean=pop_mean, alternative="two-sided"
)

print("\n=== 4. ttest_1samp() 결과 ===")
print(f"함수 반환 t통계량  : {t_stat:.4f}")
print(f"p-value            : {p_value:.4f}")

# 8. 통계량 일치 여부 비교
is_match = np.isclose(t_manual, t_stat)
print(f"t통계량 일치 여부  : {is_match}")

# 9. 귀무가설 판단
alpha = 0.05
print("\n=== 5. 가설 검정 판정 ===")
if p_value < alpha:
    print(
        f"판정: p-value({p_value:.4f}) < {alpha} 이므로 귀무가설을 기각합니다 (180,000달러와 유의한 차이가 있음)."
    )
else:
    print(
        f"판정: p-value({p_value:.4f}) >= {alpha} 이므로 귀무가설을 기각하지 못합니다 (180,000달러와 다르다고 볼 수 없음)."
    )

=== 1. 기술통계량 ===
표본 수 (n)        : 30
표본평균           : $233,323.80
표본표준편차 (s)   : $86,013.01

=== 2. 직접 계산 결과 ===
표준오차 (SE)      : $15,703.7556
직접 계산 t통계량  : 3.3956

=== 3. 정규성 검정 ===
Shapiro-Wilk 통계량: 0.9538, p-value: 0.2133
정규성 충족 여부   : 충족

=== 4. ttest_1samp() 결과 ===
함수 반환 t통계량  : 3.3956
p-value            : 0.0020
t통계량 일치 여부  : True

=== 5. 가설 검정 판정 ===
판정: p-value(0.0020) < 0.05 이므로 귀무가설을 기각합니다 (180,000달러와 유의한 차이가 있음).


### 필수 1 답변 작성란

**Q1.** t통계량은 평균 차이와 표준오차를 어떻게 이용한 값인가요?  
- 관측된 평균과 기준값의 차이를 (그 차이의 불확실성을 의미하는) 표준오차로 나눈 값

**Q2.** 같은 평균 차이라면 표준오차가 작아질수록 t통계량의 절댓값은 어떻게 변하나요?  
- 분모(표준오차)가 작아지므로 T통계랑의 절대값은 커진다.

**Q3.** 직접 계산한 t통계량과 `ttest_1samp()`의 t통계량은 일치하나요? 
- 일치한다. 두방법 모두 약 3.3956으로 계산됨

**Q4.** 검정 결과 평균 판매가격이 180,000달러와 다르다고 판단할 수 있나요?
- 그렇다. P-value가 약 0.002로 0.05보다 작아 귀무가설을 기각한다.
- 따라서 평균 판매가격은 180,000달러와 다르다.

---

## 필수 2. 독립표본 t검정의 가정 점검과 수행

### 문제 2-1. 주방 품질 `Gd`와 `TA` 집단의 평균 판매가격 비교

#### 문제 설명

주방 품질이 `Gd`인 주택과 `TA`인 주택에서 각각 20개를 추출해 평균 판매가격을 비교합니다. 두 집단은 서로 다른 주택으로 구성되어 있습니다.

#### 요구사항

1. `KitchenQual == "Gd"`와 `KitchenQual == "TA"` 집단의 `SalePrice`에서 각각 `n=20`, `random_state=42`로 표본을 추출하세요.
2. 두 집단의 표본 수와 평균을 확인하세요.
3. 데이터 수집 구조를 근거로 두 집단의 독립성을 설명하세요.
4. 각 집단에 Shapiro-Wilk 정규성 검정을 수행하세요.
5. 두 집단에 Levene 등분산 검정을 수행하세요.
6. 가정 점검 결과에 따라 `equal_var=True` 또는 `equal_var=False`를 결정하세요.
7. `stats.ttest_ind()`로 독립표본 t검정을 수행하세요.
8. t통계량과 p-value를 출력하고 두 집단 평균 차이를 해석하세요.

#### 해석 질문

**Q1.** 두 집단이 독립표본인 이유는 무엇인가요?  
**Q2.** 독립성은 별도의 p-value로 확인할 수 있나요?  
**Q3.** 정규성과 등분산성 가정은 각각 충족되나요?  
**Q4.** `equal_var`에는 어떤 값을 사용해야 하나요?  
**Q5.** 두 집단의 평균 판매가격에는 통계적으로 유의한 차이가 있나요?

#### 제출 결과

- 집단별 표본 수와 평균
- 독립성 설명
- 정규성 및 등분산성 검정 결과
- `equal_var` 선택 근거
- 독립표본 t검정 결과와 해석
- Q1~Q5 답변

In [5]:
# 필수 2 코드를 작성하세요.
import numpy as np
import pandas as pd
from scipy import stats

# 1. KitchenQual 'Gd'와 'TA' 집단에서 n=20, random_state=42 표본 추출
price_gd = (
    df[df["KitchenQual"] == "Gd"]["SalePrice"].dropna().sample(n=20, random_state=42)
)
price_ta = (
    df[df["KitchenQual"] == "TA"]["SalePrice"].dropna().sample(n=20, random_state=42)
)

# 2. 표본 수와 표본평균 확인
n_gd, n_ta = len(price_gd), len(price_ta)
mean_gd, mean_ta = price_gd.mean(), price_ta.mean()

print("=== 1. 기술통계량 ===")
print(f"KitchenQual 'Gd' (Good)            : 표본 수 = {n_gd}, 평균 = ${mean_gd:,.2f}")
print(f"KitchenQual 'TA' (Typical/Average) : 표본 수 = {n_ta}, 평균 = ${mean_ta:,.2f}")

# 4. Shapiro-Wilk 정규성 검정
stat_norm_gd, p_norm_gd = stats.shapiro(price_gd)
stat_norm_ta, p_norm_ta = stats.shapiro(price_ta)

# 5. Levene 등분산 검정
stat_levene, p_levene = stats.levene(price_gd, price_ta)

print("\n=== 2. 가정 검정 결과 ===")
print(f"Shapiro-Wilk (Gd) : 통계량 = {stat_norm_gd:.4f}, p-value = {p_norm_gd:.4f}")
print(f"Shapiro-Wilk (TA) : 통계량 = {stat_norm_ta:.4f}, p-value = {p_norm_ta:.4f}")
print(f"Levene 등분산성   : 통계량 = {stat_levene:.4f}, p-value = {p_levene:.4f}")

# 6. equal_var 결정 (Levene 검정 p-value 기준)
norm_ok = (p_norm_gd >= 0.05) and (p_norm_ta >= 0.05)
equal_var_param = bool(p_levene >= 0.05)

print("\n=== 3. 가정 판단 및 파라미터 선택 ===")
print(f"두 집단 모두 정규성 충족 여부: {norm_ok}")
print(f"등분산성 충족 여부 (p >= 0.05) : {equal_var_param}")
print(f"선택된 equal_var 파라미터     : {equal_var_param}")

# 7. 독립표본 t검정 수행
t_stat, p_value = stats.ttest_ind(price_gd, price_ta, equal_var=equal_var_param)

# 8. 결과 출력
print("\n=== 4. 독립표본 t검정 결과 ===")
print(f"t-통계량: {t_stat:.4f}")
print(f"p-value : {p_value:.4e}")

alpha = 0.05
if p_value < alpha:
    print(
        f"판정: p-value({p_value:.4e}) < {alpha} 이므로 두 집단의 평균 판매가격 차이는 통계적으로 유의합니다."
    )
else:
    print(
        f"판정: p-value({p_value:.4e}) >= {alpha} 이므로 두 집단의 평균 판매가격 차이는 통계적으로 유의하지 않습니다."
    )

=== 1. 기술통계량 ===
KitchenQual 'Gd' (Good)            : 표본 수 = 20, 평균 = $191,916.60
KitchenQual 'TA' (Typical/Average) : 표본 수 = 20, 평균 = $137,432.50

=== 2. 가정 검정 결과 ===
Shapiro-Wilk (Gd) : 통계량 = 0.9555, p-value = 0.4580
Shapiro-Wilk (TA) : 통계량 = 0.9410, p-value = 0.2503
Levene 등분산성   : 통계량 = 3.7806, p-value = 0.0593

=== 3. 가정 판단 및 파라미터 선택 ===
두 집단 모두 정규성 충족 여부: True
등분산성 충족 여부 (p >= 0.05) : True
선택된 equal_var 파라미터     : True

=== 4. 독립표본 t검정 결과 ===
t-통계량: 3.4249
p-value : 1.4891e-03
판정: p-value(1.4891e-03) < 0.05 이므로 두 집단의 평균 판매가격 차이는 통계적으로 유의합니다.


### 필수 2 답변 작성란

**Q1.** 두 집단이 독립표본인 이유는 무엇인가요?  
- GD / TA 각 집단이 서로 다른 주택으로 구성되어 하나의 관측값이 다른 관측값과 짝을 이루지 않기 때문이다.

**Q2.** 독립성은 별도의 p-value로 확인할 수 있나요?  
- 일반적으로 독립성은 별도의 검정 P-value로 확인하는 것이 아니라 데이터 수집 구조, Feature 엔지니어링 구조를 통해 판단한다.

**Q3.** 정규성과 등분산성 가정은 각각 충족되나요?  
- 필수2 문제에서는 정규성과 등분산성 위반 근거가 없다.

**Q4.** `equal_var`에는 어떤 값을 사용해야 하나요?  
- Levene p-value가 0.05보다 크므로 True를 사용한다.

**Q5.** 두 집단의 평균 판매가격에는 통계적으로 유의한 차이가 있나요?
- 유의한 차이가 있다. t검정 결과 p-value가 약 0.0015로 0.05보다 작다.
- 일반적으로 주방설계를 잘해놓으면 주택가격이 달라질 가능성이 있다.

---

## 과제. 대응표본과 독립표본 상황 구분

### 문제 3-1. 같은 주택의 보수 전후 예상 판매가격 비교

#### 문제 설명

부동산 회사가 동일한 주택 10채에 대해 보수 전 예상 판매가격과 보수 후 예상 판매가격을 각각 산정했습니다. 같은 위치의 값은 동일한 주택의 전후 가격으로 서로 짝을 이룹니다.

```python
before_price = [145, 162, 178, 155, 190, 172, 168, 181, 159, 175]
after_price  = [154, 170, 185, 164, 201, 179, 176, 190, 168, 183]
```

단위는 천 달러입니다.

#### 요구사항

1. 두 배열의 길이가 같은지 확인하세요.
2. 보수 전후 평균을 계산하세요.
3. 독립표본 t검정과 대응표본 t검정 중 적절한 방법을 선택하고 이유를 설명하세요.
4. 대응표본 t검정의 정규성 가정은 개별 배열이 아니라 `after_price - before_price` 차이값에 적용된다는 점을 확인하세요.
5. 차이값에 Shapiro-Wilk 정규성 검정을 수행하세요.
6. 적절한 t검정을 실행하고 t통계량과 p-value를 출력하세요.
7. 보수 전후 평균 예상 판매가격에 유의한 차이가 있는지 결론을 작성하세요.

#### 해석 질문

**Q1.** 이 데이터는 독립표본인가요, 대응표본인가요?  
**Q2.** 대응표본 t검정에서는 무엇의 정규성을 확인해야 하나요?  
**Q3.** `stats.ttest_ind()`가 아니라 어떤 함수를 사용해야 하나요?  
**Q4.** 검정 결과 보수 전후 예상 판매가격에 유의한 차이가 있나요?

#### 제출 결과

- 표본 관계 판단과 근거
- 전후 평균과 차이값
- 차이값의 정규성 결과
- 대응표본 t검정 결과
- 결과 해석
- Q1~Q4 답변

In [6]:
# 과제 코드를 작성하세요.
import numpy as np
import pandas as pd
from scipy import stats

# 데이터 입력
before_price = np.array([145, 162, 178, 155, 190, 172, 168, 181, 159, 175])
after_price = np.array([154, 170, 185, 164, 201, 179, 176, 190, 168, 183])

# 1. 두 배열의 길이 일치 확인
assert len(before_price) == len(
    after_price
), "두 배열의 길이가 일치하지 않습니다."
n = len(before_price)

# 2. 보수 전후 평균 계산 및 차이값 생성
mean_before = np.mean(before_price)
mean_after = np.mean(after_price)
diff = after_price - before_price
mean_diff = np.mean(diff)

print("=== 1. 기술통계량 및 차이값 ===")
print(f"표본 수 (n)       : {n}채")
print(f"보수 전 평균 가격 : ${mean_before:.2f}k (${mean_before*1000:,.0f})")
print(f"보수 후 평균 가격 : ${mean_after:.2f}k (${mean_after*1000:,.0f})")
print(f"개별 차이값 (전후): {diff.tolist()}")
print(f"평균 차이값       : +${mean_diff:.2f}k (+${mean_diff*1000:,.0f})")

# 5. 차이값(diff)의 Shapiro-Wilk 정규성 검정
stat_shapiro, p_shapiro = stats.shapiro(diff)
print("\n=== 2. 차이값 정규성 검정 (Shapiro-Wilk) ===")
print(f"W-통계량: {stat_shapiro:.4f}")
print(f"p-value : {p_shapiro:.4f} (정규성 충족 여부: {p_shapiro >= 0.05})")

# 6. 대응표본 t검정 수행 (stats.ttest_rel)
t_stat, p_val = stats.ttest_rel(after_price, before_price)
print("\n=== 3. 대응표본 t검정 결과 (ttest_rel) ===")
print(f"t-통계량: {t_stat:.4f}")
print(f"p-value : {p_val:.4e}")

# 7. 결론 도출
alpha = 0.05
if p_val < alpha:
    print(
        f"\n판정: p-value({p_val:.4e}) < {alpha} 이므로 귀무가설을 기각합니다."
    )
    print("      보수 전후 예상 판매가격에는 통계적으로 유의미한 차이가 있습니다.")
else:
    print(
        f"\n판정: p-value({p_val:.4e}) >= {alpha} 이므로 귀무가설을 기각하지 못합니다."
    )

=== 1. 기술통계량 및 차이값 ===
표본 수 (n)       : 10채
보수 전 평균 가격 : $168.50k ($168,500)
보수 후 평균 가격 : $177.00k ($177,000)
개별 차이값 (전후): [9, 8, 7, 9, 11, 7, 8, 9, 9, 8]
평균 차이값       : +$8.50k (+$8,500)

=== 2. 차이값 정규성 검정 (Shapiro-Wilk) ===
W-통계량: 0.8871
p-value : 0.1574 (정규성 충족 여부: True)

=== 3. 대응표본 t검정 결과 (ttest_rel) ===
t-통계량: 22.8079
p-value : 2.8420e-09

판정: p-value(2.8420e-09) < 0.05 이므로 귀무가설을 기각합니다.
      보수 전후 예상 판매가격에는 통계적으로 유의미한 차이가 있습니다.


### 과제 답변 작성란

**Q1.** 이 데이터는 독립표본인가요, 대응표본인가요?  
- 대응표본: 동일한 주택에 대해 보수 전 후를 비교했기 때문

**Q2.** 대응표본 t검정에서는 무엇의 정규성을 확인해야 하나요?  
- 두 표본의 차이값의 정규성을 확인해야한다.

**Q3.** `stats.ttest_ind()`가 아니라 어떤 함수를 사용해야 하나요?  
- stats.ttest_rel()

**Q4.** 검정 결과 보수 전후 예상 판매가격에 유의한 차이가 있나요?
- 네. p-value가 유의수준보다 매우 낮으므로 통계적으로 유의미하다.


## 실습 마무리

1. t통계량은 어떤 두 값을 비교하여 계산하나요?
- 관측된 평균 차이를 표준오차(그 차이의 불확실성)와 비교한다.

2. 단일표본 t검정은 어떤 상황에서 사용하나요?
- 하나의 표본평균을 특정 기준값 또는 모집단 평균과 비교할 때 사용한다.

3. 독립표본과 대응표본은 데이터 수집 구조에서 어떤 차이가 있나요?
- 독립표본은 서로 다른 대상들로 구성되며, 대응표본은 같은 대상의 반복 측정값(실험, 공사 전/후 등)처럼 각 관측값이 짝을 이룬다.

4. 독립표본 t검정에서 정규성·등분산성·독립성은 각각 어떻게 확인하나요?
- Shapiro, Levene, 독립성은 데이터 수집 구조

5. 분산이 같다고 보기 어렵거나 확신할 수 없을 때 어떤 t검정을 사용할 수 있나요?
- equal_var = False인 Welch t 검정을 사용할 수 있다.